### DISCUSSION TITLE
Testing support

### DISCUSSION THREAD

**Discussion by roelofr** [westacks/telebot#27] [Ideas]
Hey there,

I just discovered this package via my GitHub homepage (I didn't know it did that).

It looks great, but I'm seeing no docs on testing <ins>from an end-user perspective</ins> (which is something a *lot* of Telegram bot packages don't have).

What would be the best way to test this, I'd be glad to help.

**Comment by punyflash** [comment]
Hello. What exactly you mean about testing? You can't use this library for unit testing as it works as API-wrapper. It just give you autocompletion and different ways to organize your code.

If you want to run bot locally, you should use [long polling](https://westacks.github.io/telebot/#/laravel?id=long-polling). In this way you can work with your bot without configuring a webhook and then test how it works directly in messanger.

**Comment by roelofr** [comment]
Yeah, true, but since this library includes commands, I'd like to be able to make assertions on the output of commands.

So say I'd have a `/start` command, It would be most awesome if there would be helpers like these:

```php
$manager = new BotManager(…);
$manager->test('/start')->assertSendMessage(['text' => 'I can do a lot of things']);
```

This way, we can test the outputs using the helpers, so `sendMessage` would be translated to `assertSendMessage` and the params would be matched against the would-be request (so for the example above, only the `text` would be validated, and params such as `parse_mode` won't be).

**Reply by punyflash** [reply]
This is actually an interesting idea, but I would improve it a bit:
* The bot can execute multiple handlers and API-methods on each update, which should be checked as well.
* Each action can rely on (and change) the database state and custom logic from your application, so developers will need to prepare it by themselves.
* Tests should be more universal for any type of update, not only for commands, as you will not be able to reproduce user flow fully. And this is actualy the main problem, as [update type](https://core.telegram.org/bots/api#update) is quite complex. Mocking up update every time will be "pain in the *ss" for someone who write this tests.

Also I think each bot should be tested separately, so we need perform tests on `TeleBot` instance, but not `BotManager`.

So in perfect way, the testing system should lock like this:
```php
$bot = new TeleBot(...);
$update = new Update(...);

$methods = []; // Array of actions that bot performed with API. If result action is not present in this array, assertion should fail.
$handlers = []; // Array of handlers from your app that bot executed. Works in a same way as methods.
// May be there will be a room for other type of tests but for now I can only imagine only these

$test = $bot->test($update);

// Example for PHPUnit
$this->assertTrue($test->methodsExecuted($methods));
$this->assertTrue($test->handlersExecuted($handlers));
```

I do believe that this is possible to do by storing actions in some kind of testing class object instead of executing them immediately. The main focus should be on simplifying the mock up of the `Update` objects by some kind of generator class, but I don't realy imagine a simple way to do so right now, I'll definitely will try to implemet this system after v2 release or gladly help if someone start it earlier.

**Reply by punyflash** [reply]
Oh, also I missed that bot supports an anonymous functions as handlers, so this will not be comfortable to check them. May be handlers test is not realy usefull as we can't validate it depending on how developers organise their code and we should only focus on API-methods.

**Reply by roelofr** [reply]
I think this library could supply the means of testing, but the responsibility for a "testable architecture" remains with the end-developers.

I really like the idea of testing an Update through the system. It would also take away some edge cases I thought up such as identifying chats and the like. 

We could supply a `TestingFactory` of sorts, with methods like `TestingFactory::createChatMessage(string $text, Chat $chat, User $user)` to ease the update-creation-process a bit.
